### PR TITLE
refactor(lib): new `joinSQLFragments` util

### PR DESCRIPTION
**Pull request by papb**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [X] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [X] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

This is a refactor which adds the utility method `joinSQLFragments`, which enables a better way of joining SQL fragments together, also improving readability wherever it is used.

The most advantage of this new utility method is that it automatically skips falsy inputs, so we no longer need that verbose stuff for gluing together certain SQL fragments that may or may not exist.

This also prevents multiple spacing (aesthetic only) and makes code more readable since we no longer need to worry about leading or trailing spaces to ensure the SQL will be properly concatenated.

### CODE REVIEW

**Comment by papb** [comment]
Note: there are many more places in the code in which this new utility can be applied, I only refactored part of the code. Later, on another PR, I shall refactor the rest. (I didn't want to make only one single huge PR)

**Comment by sushantdhiman** [comment]
MSSQL assertions are failing, please fix them

**Comment by papb** [comment]
@sushantdhiman Whoops I missed that. By the way while doing this refactor I noticed that some queries have a trailing `;` while others don't. I wanted to remove all those, do you think this could cause problems?

File: test/unit/dialects/mariadb/query-generator.test.js [hunk 1]
```diff
@@ -763,7 +763,7 @@ if (dialect === 'mariadb') {
...
+          expectation: "SELECT CONSTRAINT_NAME as constraint_name,CONSTRAINT_NAME as constraintName,CONSTRAINT_SCHEMA as constraintSchema,CONSTRAINT_SCHEMA as constraintCatalog,TABLE_NAME as tableName,TABLE_SCHEMA as tableSchema,TABLE_SCHEMA as tableCatalog,COLUMN_NAME as columnName,REFERENCED_TABLE_SCHEMA as referencedTableSchema,REFERENCED_TABLE_SCHEMA as referencedTableCatalog,REFERENCED_TABLE_NAME as referencedTableName,REFERENCED_COLUMN_NAME as referencedColumnName FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE WHERE ( REFERENCED_TABLE_NAME = 'User' AND REFERENCED_COLUMN_NAME = 'email' ) OR ( TABLE_NAME = 'User' AND COLUMN_NAME = 'email' AND REFERENCED_TABLE_NAME IS NOT NULL )"
```
Comment (sushantdhiman) [inline test/unit/dialects/mariadb/query-generator.test.js]: ^

File: test/unit/dialects/mssql/query-generator.test.js [hunk 2]
```diff
@@ -41,7 +41,7 @@ if (current.dialect.name === 'mssql') {
...
-        mssql: "IF OBJECT_ID('[myTable]', 'U') IS NULL CREATE TABLE [myTable] ([int] INTEGER, [varchar] VARCHAR(50) UNIQUE); EXEC sp_addextendedproperty @name = N'MS_Description', @value = N'Foo Bar', @level0type = N'Schema', @level0name = 'dbo', @level1type = N'Table', @level1name = [myTable], @level2type = N'Column', @level2name = [int]; EXEC sp_addextendedproperty @name = N'MS_Description', @value = N'Bar Foo', @level0type = N'Schema', @level0name = 'dbo', @level1type = N'Table', @level1name = [myTable], @level2type = N'Column', @level2name = [varchar];" });
+        mssql: "IF OBJECT_ID('[myTable]', 'U') IS NULL CREATE TABLE [myTable] ([int] INTEGER, [varchar] VARCHAR(50) UNIQUE) ; EXEC sp_addextendedproperty @name = N'MS_Description', @value = N'Foo Bar', @level0type = N'Schema', @level0name = 'dbo', @level1type = N'Table', @level1name = [myTable], @level2type = N'Column', @level2name = [int]; EXEC sp_addextendedproperty @name = N'MS_Description', @value = N'Bar Foo', @level0type = N'Schema', @level0name = 'dbo', @level1type = N'Table', @level1name = [myTable], @level2type = N'Column', @level2name = [varchar];" });
```
Comment (sushantdhiman) [inline test/unit/dialects/mssql/query-generator.test.js]: ^

**Comment by papb** [comment]
> There is inconsistency between usage of `;`, but for now keep them as they are. Query generator methods may be mixed with others to generate complex queries, so we might break them accidentally.

Ok.

> I am noticing some extra spaces in assertions, I pointed out some of them in comments. Can we fix them?

Fixed!

Tests are passing now (except for the infamous flaky *Expected 0 running queries. 1 queries still running*)

File: lib/utils/join-sql-fragments.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ * - Strings ending with '(', which do not get a trailing space.
+ * 
+ * @param {string[]} parts
+ * @returns {string}
+ * @private
+ */
+function singleSpaceJoinHelper(parts) {
+  let result = '';
+  let skipNextLeadingSpace = true;
+  for (const part of parts) {
```
Comment (sushantdhiman) [inline lib/utils/join-sql-fragments.js]: This can be simplified using `parts.reduce`

Comment (papb) [reply lib/utils/join-sql-fragments.js]: Hmm, to be honest I don't see a way to use `parts.reduce` here in a way that could benefit readability.

The only way to use reduce here that comes to my mind is like this:

```js
let skipNextLeadingSpace = true;
return parts.reduce((acc, part) => {
  let result;
  if (skipNextLeadingSpace || doesNotWantLeadingSpace(part)) {
    result = acc + part.trim();
  } else {
    result = acc + ` ${part.trim()}`;
  }
  skipNextLeadingSpace = doesNotWantTrailingSpace(part);
  return result;
}, '');
```

Do you have another suggestion? :)

Comment (sushantdhiman) [reply lib/utils/join-sql-fragments.js]: You can include both `skipNextLeadingSpace` and `result` in accumulator (as object). Then do  `parts.reduce(() => {}).result` without using `skipNextLeadingSpace` in outer closure

Comment (papb) [reply lib/utils/join-sql-fragments.js]: @sushantdhiman Hmmm, to be honest I don't think that improves readability, but I did it anyway :)

**Pull request merged**